Document/0x05d-Testing-Data-Storage.md

```diff
@@ -1151,7 +1167,105 @@ To quickly discover potential sensitive data in the _.hprof_ file, it is also us
 
 #### Remediation
 
-In Java memory cannot be directly overwritten, instead the garbage collector will collect the object once no references are available anymore. To achieve this the object should be _nulled_ immediately after usage to reduce the attack surface.
+In Java, no immutable structures should be used to carry secrets (E.g. `String`, `BigInteger`). Nullifying them will not be effective: the Garbage collector might collect them, but they might remain in the JVMs heap for a longer period of time. 
+Rather use byte-arrays (`byte[]`) or char-arrays (`char[]`) which are cleaned after the operations are done:
+
+
+```java
+byte[] secret = null;
+try{
+	//get or generate the secret, do work with it, make sure you make no local copies
+} finally {
+	if(null != secret && secret.length > 0) {
+		for (int i = 0; i < secret; i++) {
+        array[i] = (byte) 0;
+       }
+	}
+}
+```
+
+Keys should be handled by the `AndroidKeyStore` or the `SecretKey` class needs to be adjusted. For a better implementation of the `SecretKey` one can use the class below:
+
+
+```java
+public class SecretKey implements Serializable {
```

Comment:
Since there's a `javax.crypto.SecretKey`, maybe it would make sense to change the name of this class to avoid confusion.